blockchain/chain_sync/src/network_context.rs

```diff
@@ -129,28 +134,19 @@ impl SyncNetworkContext {
         &mut self,
         peer_id: PeerId,
         request: RPCRequest,
-    ) -> Result<RPCResponse, &'static str> {
+    ) -> OneShotReceiver<RPCResponse> {
         let request_id = self.request_id;
         self.request_id.0 += 1;
+
+        let (tx, rx) = oneshot_channel();
+        self.request_table.lock().await.insert(request_id, tx);
         self.network_send
```

Comment:
Why not just pass in the return channel into the RPC message sent into the networking service so you don't have to handle the response in this circular way? Instead of sending the request, modifying a shared hashmap through a mutex, and then manually getting the inbound in another place to await the response, you just pass in the send channel and await the response with a timeout